File: sdk/typescript/src/registry.ts
/**
 * Registry client — register, search, and deregister agents
 */
import { Manifest, SearchParams, SearchResponse } from './types';

export class RegistryClient {
  constructor(private baseUrl: string) {}

  /** Register an agent manifest with the registry */
  async register(manifest: Manifest): Promise<{ id: string }> {
    const res = await fetch(`${this.baseUrl}/v1/agents`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(manifest),
    });
    if (!res.ok) throw new Error(`Registry register failed: ${res.status} ${await res.text()}`);
    return res.json() as Promise<{ id: string }>;
  }

  /** Search for agents by capability, tags, etc. */
  async search(params: SearchParams): Promise<SearchResponse> {
    const qs = new URLSearchParams();
    if (params.capability) qs.set('capability', params.capability);
    if (params.tags) qs.set('tags', params.tags.join(','));
    if (params.maxPrice !== undefined) qs.set('maxPrice', String(params.maxPrice));
    if (params.minTrust !== undefined) qs.set('minTrust', String(params.minTrust));
    if (params.available !== undefined) qs.set('available', String(params.available));
    if (params.operator) qs.set('operator', params.operator);

    const res = await fetch(`${this.baseUrl}/v1/agents/search?${qs}`);
    if (!res.ok) throw new Error(`Registry search failed: ${res.status}`);
    return res.json() as Promise<SearchResponse>;
  }

  /** Get a specific agent's manifest */
  async get(agentId: string): Promise<Manifest> {
    const res = await fetch(`${this.baseUrl}/v1/agents/${agentId}`);
    if (!res.ok) throw new Error(`Registry get failed: ${res.status}`);
    return res.json() as Promise<Manifest>;
  }

  /** Deregister an agent */
  async deregister(agentId: string): Promise<void> {
    const res = await fetch(`${this.baseUrl}/v1/agents/${agentId}`, { method: 'DELETE' });
    if (!res.ok) throw new Error(`Registry deregister failed: ${res.status}`);
  }
}

File: sdk/python/aip/registry.py
"""Registry client"""
import aiohttp
from typing import Any
from .types import Manifest, SearchResult


class RegistryClient:
    def __init__(self, base_url: str) -> None:
        self.base_url = base_url.rstrip("/")

    async def register(self, manifest: Manifest) -> dict[str, Any]:
        async with aiohttp.ClientSession() as s:
            async with s.post(f"{self.base_url}/v1/agents", json=manifest.to_dict()) as r:
                r.raise_for_status()
                return await r.json()

    async def search(self, capability: str = "", tags: list[str] | None = None) -> list[SearchResult]:
        params: dict[str, str] = {}
        if capability: params["capability"] = capability
        if tags: params["tags"] = ",".join(tags)
        async with aiohttp.ClientSession() as s:
            async with s.get(f"{self.base_url}/v1/agents/search", params=params) as r:
                r.raise_for_status()
                data = await r.json()
                return [
                    SearchResult(
                        agent_id=x["agent"]["id"], agent_name=x["agent"]["name"],
                        capability=x.get("capability", ""), endpoint=x.get("endpoint", ""),
                        trust_score=x.get("trustScore", 0),
                    )
                    for x in data.get("results", [])
                ]

    async def get(self, agent_id: str) -> dict[str, Any]:
        async with aiohttp.ClientSession() as s:
            async with s.get(f"{self.base_url}/v1/agents/{agent_id}") as r:
                r.raise_for_status()
                return await r.json()

    async def deregister(self, agent_id: str) -> None:
        async with aiohttp.ClientSession() as s:
            async with s.delete(f"{self.base_url}/v1/agents/{agent_id}") as r:
                r.raise_for_status()
